src/snowplow/tracker.lua

```diff
@@ -209,42 +183,45 @@ end
 -- @param label string: An optional string to provide additional dimensions to the event data
 -- @param property string: An optional string describing the objector the action performed on it.
 -- @param value string: A value that you can use to provide numerical data about the user event
--- @param tstamp number: Optional time (in seconds since epoch) at which event occurred
 -- @return boolean whether event was successfully collected; and the reason for failure if not
 function Tracker:track_struct_event(category, action, label, property, value)
-  local pb = payload.new_payload_builder(self.config.encode_base64)
-  pb.add_raw("e", "se")
-  pb.add("se_ca", category, validate.is_non_empty_string)
-  pb.add("se_ac", action, validate.is_non_empty_string)
-  pb.add("se_la", label, validate.is_string_or_nil)
-  pb.add("se_pr", property, validate.is_string_or_nil)
-  pb.add("se_va", value, validate.is_number_or_nil)
+  local pb = payload.new_payload_builder(self.emitter:get_request_method(), self.config.encode_base64)
+  pb:add("e", "se")
+
+  local fields = {
+    { "se_ca", "category", category, validate.is_non_empty_string },
+    { "se_ac", "action", action, validate.is_non_empty_string },
+    { "se_la", "label", label, validate.is_string_or_nil },
+    { "se_pr", "property", property, validate.is_string_or_nil },
+    { "se_va", "value", value, validate.is_number_or_nil },
+  }
+
+  for _, field in ipairs(fields) do
+    local parameter, name, var, validator = table.unpack(field)
+    validator(name, var)
+    -- The value is allowed to be nil, but we don't want to send it as a field if so
+    if var ~= nil then
+      pb:add(parameter, var)
+    end
+  end
 
   return track(self, pb)
 end
 
 -- Sends a custom unstructured event to Snowplow.
--- @param name string: The name of the event
 -- @param properties string: The properties of the event
--- @param tstamp number: Optional time (in seconds since epoch) at which event occurred
 -- @return: boolean whether event was successfully collected; and the reason for failure if not
-function Tracker:track_unstruct_event(name, properties)
-  local pb = payload.new_payload_builder(self.config.encode_base64)
-  pb.add_raw("e", "ue")
-  pb.add("ue_na", name)
-  pb.add_props("ue_px", "ue_pr", properties)
-
+function Tracker:track_unstruct_event(properties)
+  local wrapper = {
+    schema = "iglu:com.snowplowanalytics.snowplow/unstruct_event/jsonschema/1-0-0",
+    data = properties,
+  }
+  local pb = payload.new_payload_builder(self.emitter:get_request_method(), self.config.encode_base64)
```

Comment:
This is fine but a bit of a confusing interaction between tracker and emitter – first, you ask emitter for the request method in tracker to initialize the payload and then you send the payload back to emitter. Wouldn't it be better if Payload didn't require the request method information during initialization but Emitter would ask the payload for the appropriate payload form? Not necessary to change this, it's just a minor point that I think the communication between classes here could be streamlined. I can see that there seems to be a convention in Lua to give objects all information during initialization.

Reply:
I think I prefer this - it has been implemented.